src/Infrastructure.Dapper/Repositories/UserRepository.cs

```diff
@@ -254,6 +254,48 @@ public async Task UpdateUserKeyAndEncryptedDataAsync(
     }
 
 
+    public async Task UpdateUserKeyAndEncryptedDataV2Async(
+        User user,
+        IEnumerable<UpdateEncryptedDataForKeyRotation> updateDataActions)
+    {
+        await using var connection = new SqlConnection(ConnectionString);
+        connection.Open();
+
+        await using var transaction = connection.BeginTransaction();
+        try
+        {
+            if (user.AccountRevisionDate == null)
+            {
+                user.AccountRevisionDate = user.RevisionDate;
+            }
+
+            ProtectData(user);
+            await connection.ExecuteAsync(
+                $"[{Schema}].[{Table}_Update]",
+                user,
+                transaction: transaction,
+                commandType: CommandType.StoredProcedure);
+
+            //  Update re-encrypted data
+            foreach (var action in updateDataActions)
+            {
+                await action(connection, transaction);
+            }
+            transaction.Commit();
+        }
+        catch
+        {
+            transaction.Rollback();
+            throw;
+        }
+        finally
+        {
+            connection.Close();
+            UnprotectData(user);
+        }
```

Comment:
```suggestion
```

We don't need to `UnprotectData` the user entity because this method doesn't return anything.

Reply:
The idea for this was that it is unclear if the caller expects the reference they pass in to be the same after calling. I guess since we are the only caller in rotateuseraccountkeyscommand, this should be safe. But it would be unexpected from a caller's point of view for the passed in user object to be mutated.

Reply:
In C# arguments are passed by value by default. You need to use the `ref` modifier to turn things into pass by reference.
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/method-parameters

Typically, for a repository layer, you have `Task` to indicate you will be saving the object as provided, or you return a value to indicate changes.

Reply:
This can be disregarded, followed up in slack.